### DISCUSSION TITLE
find_minima() function

### DISCUSSION THREAD

**Discussion by gipsea** [skyfielders/python-skyfield#593] [Q&A]
Good evening everybody.

Iḿ really happy to have bumped into skyfield because as an amatorial programmer I found skyfield more intuitive than other python module.

At the moment I'm trying to create a script that can create sideral and sinodic table including moon phases etc. A sort of 'The american ephemeridis' books

At the moment, I'm studing the Aspectarian or sinodic angles. For congiunction and opposition I consider to use find_minima() and find_maxima() of the function separation_from(). Following the documentation online: [https://rhodesmill.org/skyfield/searches.html](url)
when it comes to minima elongation between venus and mars it never shows 0 degrees. I bumped in all this astronomy world by chance but by logic I expect that when the observed planet pass from westward to eastward (or revers, as I said I' not an astronomer) of the observing one it always touch the 0 at least on the projection onto the celestial/ecliptic equator.

is this my limitation or find_minima() of separation_from() doesn't represent the correct way finding conjunction/opposition between 2 planets?

thanks everybody for help

ciao

**Comment by JoshPaterson** [comment]
Welcome!

The `separation_from` function returns the angle between 2 objects without projecting them onto any other plane. It makes sense that it never quite reaches 0 because that would mean that the two objects were directly overlapping each other in the sky.

I don't have any experience with aspectarian angles, but I glanced over [the wikipedia article about 'astrological aspect'](https://en.wikipedia.org/wiki/Astrological_aspect) and it seems to say that the aspect is calculated using the difference in ecliptic longitude. As you mentioned in your comment we would expect that to reach zero when two objects pass each other. Luckily this is the same way that skyfield calculates conjuction and opposition. Skyfield includes a function for calculating conjunctions and oppositions, and the documentation for it is here:
[https://rhodesmill.org/skyfield/almanac.html#opposition-and-conjunction](https://rhodesmill.org/skyfield/almanac.html#opposition-and-conjunction)

I believe that existing function finds conjunctions and oppositions with the sun, so if you are looking for conjunctions and oppositions of 2 planets with each other you might be able look at the source code for `almanac.oppositions_conjunctions` and modify it so it will work with 2 planets.

@brandon-rhodes is the maintainer of Skyfield, and he might know of a better way to do what you are looking for, but in the meantime I hope this helps!

**Reply by gipsea** [reply]
Ciao Josh,

Thanks for your reply and efforts. I'll take a look to functiins and try to bend these the way I need...if I'll b good enough;)

Cheers

**Reply by JoshPaterson** [reply]
I had some extra time today and I took a look at this problem. Here is an example script that I think does what you are looking for:
```
from skyfield.api import load
from skyfield import almanac
from skyfield.constants import pi

ts = load.timescale()
eph = load('de421.bsp')

def oppositions_conjunctions_2_bodies(ephemeris, target1, target2):
    """Build a function to find oppositions and conjunctions between target1 and target2.

    See :ref:`oppositions-conjunctions` for how to call this routine and
    interpret the results.

    """
    earth_at = ephemeris['earth'].at

    def leading_or_trailing(t):
        """Return whether the target is east or west of the Sun."""
        e = earth_at(t)
        _, slon, _ = e.observe(target1).apparent().ecliptic_latlon()
        _, tlon, _ = e.observe(target2).apparent().ecliptic_latlon()
        return ((slon.radians - tlon.radians) / pi % 2.0).astype('int8')

    leading_or_trailing.step_days = 40  # Mercury
    return leading_or_trailing


t0 = ts.utc(2019, 1, 1)
t1 = ts.utc(2021, 1, 1)
f = oppositions_conjunctions_2_bodies(eph, eph['mars'], eph['jupiter barycenter'])
t, y = almanac.find_discrete(t0, t1, f)

print(t.utc_iso())
print(y)
```
for me this produces this:
```
['2019-05-05T21:56:57Z', '2020-03-20T11:34:48Z']
[1 0]
```
The date of March 20th 2020 for a conjunction of mars and jupiter is correct according to [this news article](https://www.space.com/jupiter-mars-conjunction-skywatching-march-2020.html).

`oppositions_conjunctions_2_bodies` is a slightly changed version of `almanac.oppositions_conjunctions` that allows you to specify both objects. The [documentation](https://rhodesmill.org/skyfield/almanac.html#opposition-and-conjunction) talks about how the meaning of 0 and 1 in the `y` array can change depending on if the target bodies are inferior or superior, so depending on which ones you choose they might mean different things.

**Reply by gipsea** [reply]
Hi John,

As I posted on a separate post I'm founding some issue on other function I'm implementing so I came back to this post and double check your code to see if I can find some inspiration. At the end there is a doubt that might lead to the solution...
In your code you use almanac.find_discrete whilst in the skyfield documentation the discrete function is in the serchlib module

Is there a theoretical difference in the functions despite same name (I'll check the code asap) and when use one rather the other?
cheers

**Comment by gipsea** [comment]
Ciao Josh,

thanks for your time spent to investigate my question.

For what I know conjunction and opposition are only 2 of few aspects 2 planets can fase during the revolution. Generally the angular distance is never more and 180 degree or pi in radiant.

Your script is correct considering it uses the longitudes and it only works for oppositions and conjunction. if you want to investigate a specific aspect like trine, square etc I came across to such function:
```
def delta_long(long_uno, long_due):
    if abs(long_uno-long_due)<=180:
        result = abs(long_uno - long_due)
    else:
        result = abs (min(long_uno,long_due)+360-max(long_uno, long_due))
    return result
```
long_uno and long_due are the 2 longitude that I calculate outside the function.

with similar function and find_discrete() it is possible to investigate aspects other than opposition or conjunction

Thanks again
